core/sr-staking-primitives/src/offence.rs

```diff
@@ -0,0 +1,147 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Common traits and types that are useful for describing offences for usage in environments
+//! that use staking.
+
+use rstd::vec::Vec;
+
+use codec::{Encode, Decode};
+use sr_primitives::Perbill;
+
+use crate::SessionIndex;
+
+/// The kind of an offence, is a byte string representing some kind identifier
+/// e.g. `b"im-online:offlin"`, `b"babe:equivocatio"`
+// TODO [slashing]: Is there something better we can have here that is more natural but still
+// flexible? as you see in examples, they get cut off with long names.
+pub type Kind = [u8; 16];
+
+/// Number of times the offence of this authority was already reported in the past.
+///
+/// Note we don't buffer offence reporting so everytime we see a new offence
+/// of the same kind, we will report past authorities again.
+/// This counter keeps track of how many times the authority was already reported in the past,
+/// so that we can slash it accordingly.
+pub type OffenceCount = u32;
+
+/// A type that represents a point in time on an abstract timescale.
+///
+/// See `Offence::time_slot` for details. The only requirement is that such timescale could be
+/// represented by a single `u128` value.
+pub type TimeSlot = u128;
+
+/// A trait implemented by an offence report.
+///
+/// This trait assumes that the offence is legitimate and was validated already.
+///
+/// Examples of offences include: a BABE equivocation or a GRANDPA unjustified vote.
+pub trait Offence<Offender> {
+	/// Identifier which is unique for this kind of an offence.
+	const ID: Kind;
+
+	/// The list of all offenders involved in this incident.
+	///
+	/// The list has no duplicates, so it is rather a set.
+	fn offenders(&self) -> Vec<Offender>;
+
+	/// The session index which is used for querying the validator set for the `slash_fraction`
+	/// function.
+	fn session_index(&self) -> SessionIndex;
+
+	/// Return a validator set count at the time when the offence took place.
+	fn validator_set_count(&self) -> u32;
+
+	/// A point in time when this offence happened.
+	///
+	/// This is used for looking up offences that happened at the "same time".
+	///
+	/// The timescale is abstract and doesn't have to be the same across different implementations
+	/// of this trait. The value doesn't represent absolute timescale though since it is interpreted
+	/// along with the `session_index`. Two offences considered to happen at the same time iff
+	/// both `session_index` and `time_slot` are equal.
+	///
+	/// As an example, for GRANDPA timescale could be a round number and for BABE it could be a slot
+	/// number. Note that for BABE the round number is reset each epoch.
+	fn time_slot(&self) -> TimeSlot;
```

Comment:
```suggestion
	type TimeSlot: Copy + Clone + Eq + PartialEq;
	fn time_slot(&self) -> Self::TimeSlot;
```

Reply:
i don't think that's a great excuse for clobbering type information.

putting on an `Encode` bound is trivial, and in reality, it'll end up being much more compact and efficient by using the real types (which are exclusively 32-bit and 64-bit).

they should only be coerced to a heavy-weight anonymous type like `u128` when they actually need to be compared to each other, and in that case you'd still more likely want to require an `Into<TimeSlot>` bound on an associated type rather than forcing the caller to do an ugly conversion.